ingestion-edge/ingestion_edge/publish.py

```diff
@@ -3,13 +3,40 @@
 # file, you can obtain one at http://mozilla.org/MPL/2.0/.
 
 from .conf import ROUTE_TABLE
+from .conf import QUEUE_PATH
 from datetime import datetime
 from flask import Blueprint, request
 from google.cloud.pubsub import PublisherClient
+from persistqueue import Queue
+from persistqueue.exceptions import Empty
 from typing import Tuple
+import threading
 
 publish = Blueprint('publish', __name__)
 client = PublisherClient()
+q = Queue(QUEUE_PATH)
+flush_lock = threading.Lock()
+
+
+@publish.route("/__flush__")
+def flush_queue() -> Tuple[str, int]:
+    if q.qsize() == 0:
+        # early return on empty queue
+        return "", 204
+    tasks_done = 0
+    with flush_lock:
+        while True:
+            try:
+                item = q.get_nowait()
+            except Empty:
+                break
+            try:
+                client.publish(**item).result()
+                q.task_done()
+                tasks_done += 1
+            except Exception:
+                return str(tasks_done), 202
```

Comment:
Probably worth adding a comment here to explain that by returning without calling `q.task_done()`, we aren't updating the cursor. I assume the intention here is that the next call to __flush__ will try again to send the message.

Do we need a TODO here to investigate whether there are cases where a message could fail infinitely? Something like a particularly large message might cause a non retryable error.

Or would we rely on monitoring at that point to let us know the records aren't getting flushed and that we need to investigate?

Reply:
good point, that is the intention, but that's not what would always happen. `except Exception` is meant to be replaced with something that only catches transient exceptions, so that we don't retry indefinitely.